jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FragmentingFlusher.java

```diff
@@ -0,0 +1,205 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.FrameHandler.Configuration;
+import org.eclipse.jetty.websocket.core.OpCode;
+
+/**
+ * Used to split large data frames into multiple frames below the maxFrameSize.
+ */
+public abstract class FragmentingFlusher
+{
+    private static final Logger LOG = Log.getLogger(FragmentingFlusher.class);
+
+    private final Queue<FrameEntry> entries = new ArrayDeque<>();
+    private final IteratingCallback flusher = new Flusher();
+    private final Configuration configuration;
+    private boolean canEnqueue = true;
+
+    public FragmentingFlusher(Configuration configuration)
+    {
+        this.configuration = configuration;
+    }
+
+    abstract void forwardFrame(Frame frame, Callback callback, boolean batch);
+
+    public void sendFrame(Frame frame, Callback callback, boolean batch)
+    {
+        boolean enqueued = false;
+        synchronized (this)
+        {
+            if (canEnqueue)
+                enqueued = entries.offer(new FrameEntry(frame, callback, batch));
+        }
+
+        if (enqueued)
+            flusher.iterate();
+        else
+            notifyCallbackFailure(callback, new ClosedChannelException());
```

Comment:
shouldn't think be the original failure?  Perhaps `boolean canEnqueue` should become `Throwable failure` ?